### PR TITLE
Save bookend state in window history

### LINKED ISSUE #15095: Store BOOKEND_STATE in history state
We should store the `BOOKEND_STATE` in the history state so that if users navigate back to the page they were on, the bookend will still be open if it was before.

### PR DESCRIPTION
**Pull request by Enriqe**
Re-do of #17041 

Fixes #15095.

When users navigate to another page from the bookend and come back, the bookend is closed, which might be confusing.

This PR saves the bookend state so when a user opens it, navigates to another page and comes back, it will still be open.

### CODE REVIEW

**Comment by Enriqe** [comment]
Demo available at https://bookend-skip-animation.firebaseapp.com/examples/amp-story/artezan/bookend.html

**Comment by gmajoulet** [comment]
High level comments:

### layoutCallback
Your new code in the `layoutCallback` are going to:
1. Wait until the bookend component is build and returns its implementation
2. Wait until the bookend config is fully loaded
3. Finally call switchTo() and get the story playing and interactive.

The story will be unresponsive (can't navigate, videos won't play, progress bar won't be updated) until the bookend is built, and then until the config is loaded, which could take multiple seconds.
We should find another way of doing it that doesn't hurt this UX this bad.

### History states
If you want to provide new helpers to write/read in the browser history, you should use these for the two cases we have (activePageId, bookendState). Having `setHistoryStatePageId_`, `getHistoryStatePageId_`, `setHistoryBookendState_`, `getHistoryState_`, `updateHistoryState_`, is not a good API.
Could we simply have `setHistoryState_(key, value)` and `getHistoryState(key)` instead?


Additionally, should we preserve the scroll position inside the bookend element?

**Comment by Enriqe** [comment]
@gmajoulet we actually want to block the layoutCallback when there the user has a bookend state per last PR's discussion (https://github.com/ampproject/amphtml/pull/17041#issuecomment-408538951), although I do need to fix the PR to actually only block it when the user has the bookend state in history.

We also said conserving the `scrollTop` position might be out of scope for this PR:
https://github.com/ampproject/amphtml/pull/17041#issuecomment-408497604

I'll also do the refactoring of the helpers. Let me know if you have any other comments.

**Comment by Enriqe** [comment]
PTAL @newmuis @gmajoulet

**Comment by gmajoulet** [comment]
Cool!!

As seen offline, that code could become a lot easier if we spend some time refactoring the `pagination-buttons` to move away from the `/navigation-state.StateChangeEventDef` events, and use the store service instead.
It should remove the need for the `this.openedBookend_` property, and the need to open the bookend only after the `switchTo`.
Thanks!

**Comment by Enriqe** [comment]
Filed #19600, will update the PR after that gets fixed.

**Comment by Enriqe** [comment]
Ready for review 👍

File: extensions/amp-story/1.0/amp-story.js [hunk 1]
```diff
@@ -1589,8 +1581,39 @@ export class AmpStory extends AMP.BaseElement {
...
     this.element.classList.toggle('i-amphtml-story-bookend-active', isActive);
+    this.updateHistoryState_(HistoryStates.BOOKEND_ACTIVE, isActive);
+  }
+
+  /**
+   * Updates the value for a given state in the window history.
+   * @param {string} stateName
+   * @param {string|boolean} newValue
+   */
+  updateHistoryState_(stateName, newValue) {
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Nit: `setHistoryState_(stateName, value)` so we have a super simple `get/set` API. We could be setting a history state for the first time.

File: extensions/amp-story/1.0/amp-story.js [hunk 2]
```diff
@@ -1589,8 +1581,39 @@ export class AmpStory extends AMP.BaseElement {
...
+    const {history} = this.win;
+    if (history.replaceState && getState(history) &&
+        this.getHistoryState_(stateName) !== newValue) {
+      getState(history)[stateName] = newValue;
+      history.replaceState(getState(history), '');
+    } else if (!getState(history)) {
+      // First time writing to history state. Nothing to overwrite.
+      history.replaceState({[stateName]: newValue}, '');
+    }
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Could we simply do:
```
const newHistory = Object.assign({}, getState(history) || {}, {[stateName]: value});
history.replaceState(newHistory);
```

**Comment by Enriqe** [comment]
PTAL. You can play around with it here:

https://bookend-skip-animation.firebaseapp.com/examples/amp-story/artezan/bookend.html

File: extensions/amp-story/1.0/amp-story.js [hunk 3]
```diff
@@ -1642,8 +1634,36 @@ export class AmpStory extends AMP.BaseElement {
...
     this.toggleElementsOnBookend_(/* display */ isActive);
     this.element.classList.toggle('i-amphtml-story-bookend-active', isActive);
+    this.setHistoryState_(HistoryStates.BOOKEND_ACTIVE, isActive);
+  }
+
+  /**
+   * Updates the value for a given state in the window history.
+   * @param {string} stateName
+   * @param {string|boolean} value
+   */
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: `@private`

File: extensions/amp-story/1.0/amp-story.js [hunk 4]
```diff
@@ -1642,8 +1634,36 @@ export class AmpStory extends AMP.BaseElement {
...
+    const newHistory = Object.assign({}, /** @type {!Object} */ (state),
+        {[stateName]: value});
+
+    history.replaceState(newHistory, '');
   }
 
+  /**
+   * Returns the value of a given state of the window history.
+   * @param {string} stateName
+   */
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: `@private @return`

File: extensions/amp-story/1.0/amp-story.js [hunk 5]
```diff
@@ -1642,8 +1634,36 @@ export class AmpStory extends AMP.BaseElement {
...
+   * @param {string} stateName
+   * @param {string|boolean} value
+   */
+  setHistoryState_(stateName, value) {
+    const {history} = this.win;
+    const state = getState(history) || {};
+    const newHistory = Object.assign({}, /** @type {!Object} */ (state),
+        {[stateName]: value});
+
+    history.replaceState(newHistory, '');
```
Comment (gmajoulet) [inline extensions/amp-story/1.0/amp-story.js]: Is that second parameter needed? It looks like it's ignored by browsers for now, but if we can remove it or set it to undefined, maybe it's safer? If the linter requires it, then fine : )

Comment (Enriqe) [reply extensions/amp-story/1.0/amp-story.js]: It is needed by the compiler :| it wouldn't let me set it to undefined either.

**Pull request merged**